### PR TITLE
fix(dot/state, lib/babe, lib/trie): improve syncing between gossamer authority nodes

### PR DESCRIPTION
**Pull request by noot**
## Changes

- pause block production when syncing, fix locking mechanism during pause/resume
- update `trie.Snapshot` to return new trie, previously it was returning old trie, so both syncer and babe were accessing the same trie (as the receiver of `Snapshot` was returned by `TrieState` in both cases)
- fix `HasFinalizedBlock` logic and db key

## Tests

```
go test ./lib/trie
go test ./lib/babe
go test ./dot/state
```

## Issues

- related to #1587

### CODE REVIEW

**Review by timwu20 — APPROVED** [review]

File: lib/runtime/storage/trie.go [hunk 1]
```diff
@@ -61,7 +61,8 @@ func (s *TrieState) Snapshot() *trie.Trie {
 func (s *TrieState) BeginStorageTransaction() {
 	s.lock.Lock()
 	defer s.lock.Unlock()
-	s.oldTrie = s.t.Snapshot()
+	s.oldTrie = s.t
```
Comment (EclesioMeloJunior) [inline lib/runtime/storage/trie.go:64]: `s.oldTrie`  points to the same ref as `s.t`, if I'm not wrong, any modifications in `s.t` will modify the `s.oldTrie`

**Review by edwardmack — APPROVED** [review]
Nice work, lgtm.

File: lib/runtime/storage/trie.go (hunk 1 above)
Comment (noot) [reply lib/runtime/storage/trie.go:64]: it does, but in the following line `s.t` gets snapshotted, so `s.oldTrie` won't be modified if `s.t` is modified when this function completes. `s.oldTrie` is effectively a pointer to the root of the previous trie, and `s.t` becomes a pointer to the root of the working trie

Comment (EclesioMeloJunior) [reply lib/runtime/storage/trie.go:64]: Got it! just `s.t` changes the ref, `s.oldTrie` keeps pointing to the old, nice! Why not just keep `s.oldTrie = s.t.Snapshot` as before?

Comment (noot) [reply lib/runtime/storage/trie.go:64]: `s.t.Snapshot()` now returns the new trie to be modified, previously it returned the old trie to be saved and `s.t` would continue to be modified

**Review by EclesioMeloJunior — APPROVED** [review]

**Pull request merged**